### ISSUE TITLE
NumberKeypad: is armedForNewEntry feature needed?

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/scenery-phet#280]
@jbphet and I discussed briefly, and I now understand the intent. Calling `armForNewEntry()` will put the keypad in a state when the next keypress will clear the value.  This feature was added for Arithmetic, but it's not used anywhere.  So @jbphet is going to investigate whether it's still needed, or whether he came up with a better way of handling this in the client.

**Comment by pixelzoom** [comment]
If this feature does survive, let's consider naming it something that's more indicative of what it does. Something like `clearOnKeyNextPress`.

**Comment by pixelzoom** [comment]
There's also a little bug here... If the client provided `valueStringProperty`, then calling `valueStringProperty.reset` doesn't necessarily clear the string.  This is also true of the `clear` function.  In both places we may want to explicitly do `valueStringProperty.value = ''`.

**Comment by pixelzoom** [comment]
The code for this feature is duplicated in 2 places:

```js
      if ( self.armedForNewEntry ) {
        self.valueStringProperty.reset();
        self.armedForNewEntry = false;
      }
```

**Comment by pixelzoom** [comment]
Instead of calling `valueStringProperty.reset`, I set it to the empty string. The duplicated code is now:

```js
      if ( self.armedForNewEntry ) {
        self.valueStringProperty.value = '';
        self.armedForNewEntry = false;
      }
```

**Comment by pixelzoom** [comment]
Looks like a bug in the backspace key:

```js
 if ( self.valueStringProperty.value.length > 0 ) {

          // remove the last digit from the current digit string
          var shortenedDigitString = self.valueStringProperty.value.slice( 0, -1 );

          if ( self.armedForNewEntry ) {
            self.valueStringProperty.value = '';
            self.armedForNewEntry = false;
          }

          // set the new shortened value
          self.valueStringProperty.value = shortenedDigitString;
        }
```

Why is the `armedForNewEntry` code executing *after* the string is shortened? Shouldn't it behave like all other keys?

**Comment by jbphet** [comment]
After looking at Arithmetic, I found that this feature **is** used.  There is a method ```armForNewEntry``` that sets the ```armedForNewEntry``` flag.  So yes, this feature is still needed in order to support the case where the user starts entering a new answer and therefore needs to clear the current value on the next key press.

**Comment by jbphet** [comment]
Below are some responses to other questions posed by @pixelzoom:

> If this feature does survive, let's consider naming it something that's more indicative of what it does. Something like clearOnKeyNextPress.

I'm good with that name.

>  In both places we may want to explicitly do valueStringProperty.value = ''.

Good observation, and I'm fine with explicitly setting the string property value to a zero length string.

> Why is the ```armedForNewEntry``` code executing after the string is shortened? Shouldn't it behave like all other keys?

The backspace key intentionally behaves differently.  The rationale is that if a user has entered an incorrect value and wants to correct it, and starts hitting the backspace, it should work like the backspace always does instead of clearing the display.

**Comment by pixelzoom** [comment]
@jbphet said:
> The backspace key intentionally behaves differently. 

That should be documented (I'll paste your description into the code).  Might also want to consider making that configurable - it may have been desirable in your sims, but perhaps not in general.

@jbphet Thoughts one the duplicated code for this feature? (identified in https://github.com/phetsims/scenery-phet/issues/280#issuecomment-259010729)

**Comment by pixelzoom** [comment]
Seems like there's unnecessary/confusing work being done in the backspace listener. It's currently:

```js
      listener: function() {
        if ( self.valueStringProperty.value.length > 0 ) {

          //TODO shouldn't this be done after armedForNewEntry?
          // remove the last digit from the current digit string
          var shortenedValueString = self.valueStringProperty.value.slice( 0, -1 );

          //TODO duplicated code
          if ( self.armedForNewEntry ) {
            self.valueStringProperty.value = '';
            self.armedForNewEntry = false;
          }

          // set the new shortened value
          self.valueStringProperty.value = shortenedValueString;
        }
      }
```

Based on what you said in https://github.com/phetsims/scenery-phet/issues/280#issuecomment-259288174, isn't this sufficient?

```js
      listener: function() {
        if ( self.valueStringProperty.value.length > 0 ) {
          self.armedForNewEntry = false;
          self.valueStringProperty.value = self.valueStringProperty.value.slice( 0, -1 );
        }
      }
```

**Comment by pixelzoom** [comment]
@jbphet I reimplemented the backspace button listener as shown above. It works great in scenery-phet demo, but not in Arithmetic, and I don't understand why.

**Comment by pixelzoom** [comment]
@jbphet wrote:
> After looking at Arithmetic, I found that this feature is used. 

It's also used in make-a-ten and projectile-motion.

**Comment by pixelzoom** [comment]
Here's your bug in arithmetic.  WorkspaceNode line 182:

```js
      model.inputProperty.link( function( input ) {
        if ( model.state === GameState.DISPLAYING_INCORRECT_ANSWER_FEEDBACK ) {
          model.retryProblem();
        }
        updateCheckButtonState();
      } );
```

This is now firing when the backspace key is pressed, and `model.retryProblem()` clears the entire string.  So if you enter '22', press 'Check' button, then hit backspace, you get '' instead of '2'.  This bug didn't manifest previously because the button listener was storing the shortened string in a local variable, then clearing the string, then setting the string.

**Comment by pixelzoom** [comment]
@jbphet and I discussed the buggy behavior in Arithmetic via Skype.  It is indeed relying on the fact that `valueStringProperty`  was being set multiple times in the original implementation of the backspace button listener.  Restoring the buggy implementation of the backspace button listener is not an option. 

This buggy behavior is also present in Area Builder.  Projectile Motion looks OK.  I'll create issues for the buggy sims, and assign to @jbphet.

**Comment by pixelzoom** [comment]
As noted in https://github.com/phetsims/arithmetic/issues/178#issuecomment-259503429, I think we should delete this feature, it's a misplaced responsibility:

>  I think that this [Arithmetic] problem is a result of the dubious `armForNewEntry` feature in NumberKeypad. The responsibility for clearing the keypad is misplaced - it should be up to the client to clear it explicitly by calling NumberKeypad.clear. And if doing something special with the backspace button is important (which it obviously is in arithmetic and area-builder) then we could add the ability to listen to the backspace button.

**Comment by jbphet** [comment]
I fixed Area Builder and Arithmetic to work correctly with the simplified backspace behavior within NumberKeypad.

I'm not sold, at least not yet, on the idea of getting rid of this feature.  I thought about making the sims responsible for clearing the string if the user enters something while in the ```DISPLAYING_INCORRECT_ANSWER_FEEDBACK``` state, but the problem with that is that it means clearing the value of the string from inside the handler that is informing the client that the string has changed.  In other words, the code would be modifying the value of a property inside and change notification handler for that property, which does not seem like a good idea.  In fact, I think this use case is what motivated the feature in the first place.

So, my vote is to retain the feature.  @pixelzoom - back to you for further discussion or closure.

**Comment by pixelzoom** [comment]
Alright. But the current API only allows the client to set the flag, no way to clear it.  I've addressed that in the above commits, tested in all clients.   Usage is now:

```js
keypad.armedForNewEntry = true;
keypad.armedForNewEntry = false;
```

@jbphet Please review.

**Comment by jbphet** [comment]
I tested these changes - lightly - in Area Builder and Arithmetic and they seem to be working.

I had a couple of comments after reviewing the code:
+ There was some mention about decimal point not clearing the value if ```armedForNewEntry``` was set.  Why would the decimal point be any different?  What if a user entered a value like 0.52 and then tried to enter 0.25 by just pressing '.', '2', and '5'?
+ Weren't we thinking of renaming this to something like ```clearOnNextKeyPress``` (see https://github.com/phetsims/scenery-phet/issues/280#issuecomment-259010047)?

**Comment by jbphet** [comment]
It may be a bit late in the game for this idea, but what about making this a bit more granular by having NumberKeypad do nothing other than let the client know exactly what key was just pressed?  It could have an emitter and a ```keyPressed``` field, and clients are responsible for interpreting the digit.  Then we could have IntegerNumberKeypad and DecimalNumberKeypad and any specific behavior needed for those could be handled at that level.  That way, if some sim-specific custom behavior was needed, the developer could grab the keypad and make their own wrapper instead of trying to put the new behavior into the NumberKeypad class.

**Comment by pixelzoom** [comment]
> There was some mention about decimal point not clearing the value

Sorry, that should have been "backspace".  Doc is fixed.

> Weren't we thinking of renaming this to something like `clearOnNextKeyPress`

Could do, but I don't have time to go through this again, and I don't intend to use this feature in the foreseeable future.  If you'd like to change it, you have my support.

> what about making this a bit more granular by having  NumberKeypad do nothing other than let the client know exactly what key was just pressed

That crossed my mind when we were discussing getting rid of the `armedForNewEntry` feature and adding a listener to the backspace button.  But it sounds like a "start over" scenario, which I don't have time for.  On the other hand, it might be a good approach considering the recent design discussion about having an "equation entry keypad".   Keep it in mind as design progresses.

**Comment by jbphet** [comment]
I went ahead and made the change from `armedForNewEntry` to `clearOnNextKeyPress` because I thought it would improve future maintainability.  I tested in Area Builder, Arithmetic, and the scenery-phet demo.  Assigning to @pixelzoom for review, if it looks okay then I think this can be closed.

I'll keep in mind the idea of separating out the keypad from the logic that decides how to collect and validate a string.  I find the idea quite appealing, but I think neither @pixelzoom nor I have the time to pursue this right now.  If we **do** decide to pursue it, we should open a separate issue.
